File: GAME/BMP_conv.py
import math

class RGB:
	def __init__(self, r, g, b, file_data = None):
		self.r = r
		self.g = g
		self.b = b
		self.data = file_data

	def __add__(self, other):
		return RGB(self.r + other.r, self.g + other.g, self.b + other.b)

	def __div__(self, other):
		if not isinstance(other, int):
			raise ArithmeticError("non scalar div")
		return RGB(self.r / other, self.g / other, self.b / other)

	def dist(self, other):
		return math.sqrt(self.r * other.r + self.g * other.g + self.b * other.b)


def get_num(buf, s = False):
	res = 0
	for i in range(len(buf)):
		res |= buf[i] << (i << 3)
	return res

def get_colors(fns):
	colors = []
	for fn in fns:
		with open(fn, 'rb') as f:
			pass
	return colors
	

def get_side(colors):
	max_val = 0
	for comp in [lambda x: x.r, lambda x: x.g, lambda x: x.b]
		val = max(colors, comp) - min(colors, comp)
		if val >= max_val:
			max_val = val
			result = comp
	return result

def GLA(sets, colors):
	represes = map(sets, lambda s: sum(s) / len(s))
	tmp_sets = {}
	for color in colors:
		repres = min(map(represes, color.dist))
		if repres in sets:
			sets[repres].append(color)
		else:
			sets[repres] = [color]
	return tmp_sets.values()

def MC(color_sets):
	for i in range(255):
		cur_set = max(color_sets, len)
		color_sets.remove(cur_set)
		cur_set.sort(key = get_side(cur_set))
		color_sets.add(cur_set[:len(cur_set)/2])
		color_sets.add(cur_set[len(cur_set)/2:])


if __name__ == "__main__":
	colors = get_colors(filenames)
	sets = MC(set([colors]))
	for i in range(100):
		sets = GLA(sets, colors)
	collect_files(sets)

File: GAME/converter/bmp_conv.cpp
#include <iostream>
#include <set>
#include <vector>
#include <map>
#include <algorithm>
#include "rgb.h"

using namespace std;

void gla(vector<vector<RGB>>& sets, vector<RGB> const& colors) {
	vector<RGB> represes;
	for (auto const& s : sets) {
		RGB srgb;
		for (auto const& rgb : s) {
			srgb += rgb;
		}
		represes.push_back(srgb / s.size());
	}
	map<RGB, vector<RGB>> tmpSets;
	for (auto const& color : colors) {
		uint32_t dist = -1;
		RGB repres;
		for (auto const& rep : represes) {
			uint32_t d = color.dist(rep);
			if (dist > d) {
				dist = d;
				repres = rep;
			}
		}
		if (tmpSets.contains(repres)) {
			tmpSets[repres].push_back(color);
		} else {
			tmpSets[repres] = vector<RGB>{color};
		}
	}
	sets.clear();
	for (auto const& s : tmpSets) {
		sets.push_back(s.second);
	}
}

void medianCut(set<vector<RGB>>& colorSets) {
	auto r = [](RGB const& a, RGB const& b) {return a.r < b.r;};
	auto g = [](RGB const& a, RGB const& b) {return a.g < b.g;};
	auto b = [](RGB const& a, RGB const& b) {return a.b < b.b;};

	for (int i = 0; i < 255; i++) {
		vector<RGB>& curSet = max(colorSets, [](auto const& a, auto const& b){return a.size() < b.size();});
		colorSets.erase(curSet);
		uint32_t m = max(curSet, r) - min(curSet, r);
		auto comp = r;
		for (auto f : {g, b}) {
			auto s = max(curSet, f) - min(curSet, f);
			if (m > s) {
				m = s;
				comp = f;
			}
		}
		sort(curSet.begin(), curSet.end(), comp);
		m = curSet.size() / 2;
		colorSets.insert(vector<RGB>(curSet.begin(), curSet.end() - m));
		colorSets.insert(vector<RGB>(curSet.begin() + m, curSet.end()));
	}
}


int main(int n, char** args) {
	
}
